### PR TITLE
Add ArgumentOutOfRangeException.ThrowIf{Not}Equal

### LINKED ISSUE #82667: [API Proposal]: ArgumentOutOfRangeException.ThrowIfEqual and ThrowIfNotEqual
### Background and motivation

As #69590 added new throw helpers for `ArgumentOutOfRangeException`, I've started playing around with them trying to move an existing codebase to these new APIs, and one thing that quickly became apparent was how the "equal/not equal" case was currently missing. This proposal is about adding such APIs as well, so that these scenarios can also be supported.

### API Proposal

```csharp
namespace System;

public class ArgumentOutOfRangeException
{
    public static void ThrowIfEqual<T>(T value, T other, [CallerArgumentExpression("value")] string? paramName = null) where T : IEquatable<T>;
    public static void ThrowIfNotEqual<T>(T value, T other, [CallerArgumentExpression("value")] string? paramName = null) where T : IEquatable<T>;
}
```

I've tried to mirror the design of existing APIs such as `ArgumentOutOfRangeException.ThrowIfGreaterThanOrEqual`, which also take a non-nullable generic, constrained to an interface (in that case it's `IComparable<T>`).

### API Usage

Eg. in ComputeSharp I have a bunch of methods to copy to/from MD textures. Consider this one, using this API:

```csharp
public static void CopyTo<T>(this Texture2D<T> source, T[,] destination)
    where T : unmanaged
{
    ArgumentNullException.ThrowIfNull(source);
    ArgumentNullException.ThrowIfNull(destination);
    ArgumentOutOfRangeException.ThrowIfNotEqual(destination.GetLength(0), source.Height, nameof(destination));
    ArgumentOutOfRangeException.ThrowIfNotEqual(destination.GetLength(1), source.Width, nameof(destination));

    source.CopyTo(ref destination[0, 0], destination.Length, 0, 0, source.Width, source.Height);
}
```

In general this would be useful whenever people need to explicitly check against a single value (same as eg. `ThrowIfZero`).

### Open questions

The `IComparable<T>` constraint means `Enum` values can't use this API. Should we not use the constraint and instead just have the API call `EqualityComparer<T>.Default.Equals` instead? Are there performance considerations in that case?

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/82667

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/ArgumentOutOfRangeException.cs [hunk 1]
```diff
@@ -161,6 +156,26 @@ public static void ThrowIfNegativeOrZero<T>(T value, [CallerArgumentExpression(n
...
+        /// <param name="other">The value to compare with <paramref name="value"/>.</param>
+        /// <param name="paramName">The name of the parameter with which <paramref name="value"/> corresponds.</param>
+        public static void ThrowIfEqual<T>(T value, T other, [CallerArgumentExpression(nameof(value))] string? paramName = null) where T : IEquatable<T>?
+        {
+            if (EqualityComparer<T>.Default.Equals(value, other))
+                ThrowEqual(paramName, value, other);
+        }
+
+        /// <summary>Throws an <see cref="ArgumentOutOfRangeException"/> if <paramref name="value"/> is not equal to <paramref name="other"/>.</summary>
+        /// <param name="value">The argument to validate as equal to <paramref name="other"/>.</param>
```
Comment (buyaa-n) [inline src/libraries/System.Private.CoreLib/src/System/ArgumentOutOfRangeException.cs:170]: NIT: Not sure if it was on purpose, but it seems `equal` and `not equal` swapped between 2
```suggestion
        /// <param name="value">The argument to validate as equal to <paramref name="other"/>.</param>
        /// <param name="other">The value to compare with <paramref name="value"/>.</param>
        /// <param name="paramName">The name of the parameter with which <paramref name="value"/> corresponds.</param>
        public static void ThrowIfEqual<T>(T value, T other, [CallerArgumentExpression(nameof(value))] string? paramName = null) where T : IEquatable<T>?
        {
            if (EqualityComparer<T>.Default.Equals(value, other))
                ThrowEqual(paramName, value, other);
        }

        /// <summary>Throws an <see cref="ArgumentOutOfRangeException"/> if <paramref name="value"/> is not equal to <paramref name="other"/>.</summary>
        /// <param name="value">The argument to validate as not equal to <paramref name="other"/>.</param>
```

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/ArgumentOutOfRangeException.cs:170]: ThrowIfEqual throws an exception of they're equal and thus serves to validate that they're not equal.

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 2]
```diff
@@ -4,6 +4,8 @@
 // Changes to this file must follow the https://aka.ms/api-review process.
 // ------------------------------------------------------------------------------
 
+using System.Runtime.CompilerServices;
+
```
Comment (buyaa-n) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:8]: NIT:
```suggestion
```

**Review by buyaa-n — APPROVED** [review]
Overall LGTM

**Comment by stephentoub** [comment]
Thanks.

**Review by jeffhandley — APPROVED** [review]

**Comment by xtqqczze** [comment]
@stephentoub I've proposed `ArgumentOutOfRangeException.ThrowIfNotZero` (#84415).

**Pull request merged**